### PR TITLE
Async typeahead

### PR DESCRIPTION
**Pull request by roblevintennis**
**Description:**
- [Asana Task](https://app.asana.com/0/1150068311310825/1146751046397450/f)
- async search typeahead

**Screenshots:**

Here's first stab at pulling this in to [Nora (PR)](https://github.com/getethos/ethos/pull/2167)
image

- [ ] TODO: Add `entitiesKey` so we can index into the `data.items` as Nora API has it (so we'd pass a prop `entitiesKey: "items"` in this case. See my detailed [comment in other PR](https://github.com/getethos/ethos/pull/2167/files#r372997542)

### CODE REVIEW

File: src/components/AsyncTypeahead/AsyncTypeahead.js [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  const {
+    currentActive,
+    itemIndex,
+    item,
+    dataKey,
+    selectedOption,
+    setSelectedOptionDelegate,
+    onChange,
+  } = props
+  useImperativeHandle(forwardedRef, () => ({
```
Comment (stephaniel96) [inline src/components/AsyncTypeahead/AsyncTypeahead.js]: Ooooooo never used this before. Exciting!

File: src/components/AsyncTypeahead/AsyncTypeahead.js [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  const setSelectedOptionDelegate = (index) => {
+    setSelectedOption(index)
+    setActiveOption(index)
+  }
+
+  /**
+   * Determines if the user's input value is shorter then the
+   * currently selected value object.
+   */
+  const termIsLessThenValue = (inputValue) => {
```
Comment (stephaniel96) [inline src/components/AsyncTypeahead/AsyncTypeahead.js]: I'm having a bit of a hard time wrapping my head around this logic and how it's used to clear the search.

What does `term` and value refer to in this case?

File: src/components/index.tests.tsx [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react'
+import { AsyncTypeahead } from './index'
```
[the remark this answers is not in the document]
Comment (roblevintennis) [reply src/components/index.tests.tsx]: Ah good question. So, if we want to be able to import from consumer without having to know the paths to every component we can export them from the top level `index.js` and then just destructure them out (like we do now):

`import { foo, bar, baz } from 'ethos-design-system`

vs.

`import { foo } from 'ethos-design-system/path/to/foo`

and so on.

File: src/components/fa.js [hunk 4]
```diff
@@ -11,5 +11,5 @@ import { library } from '@fortawesome/fontawesome-svg-core'
 
 // We need to add any icons we want to use here on an as needed basis. This will
 // hopefully help to keep the footprint small, and is the recommended approach.
-import { faSearch } from '@fortawesome/pro-regular-svg-icons'
-library.add(faSearch)
+import { faSearch, faCircleNotch } from '@fortawesome/pro-regular-svg-icons'
+library.add(faSearch, faCircleNotch)
```
[the remark this answers is not in the document]
Comment (roblevintennis) [reply src/components/fa.js]: Monorepo / funnel is no where near able to do tree-shaking I believe being on Webpack 3 and having other obstacles. That said, perhaps Nora will be able to do this. 

And I think your article is incredibly helpful thanks and I will switch to the longer form paths individually!

I think examples are best for certain topics like tree shaking which can be confusing, so I'd encourage you do look at a very simple experiment I did last May:

https://github.com/roblevintennis/Treeshaking-Experiment

[the remark this answers is not in the document]
Comment (roblevintennis) [reply src/components/fa.js]: 326482f addresses this

File: src/components/SearchInput/SearchInput.js [hunk 5]
```diff
@@ -4,35 +4,52 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
...
-      onEnter(value)
-    }
-  }
-
-  const onChange = (ev) => {
+  const [lastValue, setLastValue] = useState('')
+  const handleOnChange = (ev) => {
     const val = ev.target.value
-    setValue(val)
+    console.log('SearchInput --> handleOnChange: ', val)
```
[the remark this answers is not in the document]
Comment (roblevintennis) [reply src/components/SearchInput/SearchInput.js]: 280858a 👍

File: src/components/SearchInput/SearchInput.js [hunk 6]
```diff
@@ -4,35 +4,52 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
...
-  const onKeyDown = (ev) => {
-    if (['Enter'].includes(ev.key) || ev.keyCode === 27) {
-      ev.preventDefault()
-      onEnter(value)
-    }
-  }
-
-  const onChange = (ev) => {
+  const [lastValue, setLastValue] = useState('')
+  const handleOnChange = (ev) => {
```
[the remark this answers is not in the document]
Comment (roblevintennis) [reply src/components/SearchInput/SearchInput.js]: Just sort of react controlled component boilerplate

File: src/components/AsyncTypeahead/useFetchEntities.test.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    await act(async () => {
+      testHook(() => {
+        const fetchCallback = jest.fn()
+        const { entities, loading } = useFetchEntities({
+          searchString: 'abc',
+          fetchEntities: fetchCallback,
+          delay: 300,
+        })
+        expect(entities).toBeDefined()
+        expect(loading).toBeDefined()
```
[the remark this answers is not in the document]
Comment (roblevintennis) [reply src/components/AsyncTypeahead/useFetchEntities.test.js]: yeah, this just sort of confirms the hook is properly exporting destructurable api promise in:

`  return { entities, loading }`

Not a super thorough test but does confirm it exposes the promised return struct

File: src/components/AsyncTypeahead/AsyncTypeahead.js (hunk 2 above)
Comment (roblevintennis) [reply src/components/AsyncTypeahead/AsyncTypeahead.js]: Maybe a bad name, I meant it as in "search term is less then current value". I suppose value is probably getting overloaded too much 🤔 my comment is actually telling what's going on. If I have time I'll contemplate some better variable names to make this less confusing

File: src/components/AsyncTypeahead/useFetchEntities.test.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (stephaniel96) [reply src/components/AsyncTypeahead/useFetchEntities.test.js]: got it!

File: src/components/AsyncTypeahead/AsyncTypeahead.js (hunk 2 above)
Comment (roblevintennis) [reply src/components/AsyncTypeahead/AsyncTypeahead.js]: hopefully fixed in latest...thanks for feedback!

**Pull request merged**